feat: New connection management docs (#29166)
supabase/supabase: apps/docs/content/guides/database/connection-management.mdx

---
id: 'connection-management'
title: 'Connection management'
description: 'Managing connections'
subtitle: 'Using your connections resourcefully'
---

## Connections

Every [Compute Add-On](/docs/guides/platform/compute-add-ons) has a pre-configured direct connection count and Supavisor pool size. This guide discusses ways to observe and manage them resourcefully.

### Configuring Supavisor's pool size

> You can change how many database connections Supavisor can manage by altering the pool size in the [Database Settings](/dashboard/project/_/settings/database):

Reviewer: ```suggestion
You can change how many database connections Supavisor can manage by altering the pool size in the "Connection pooling configuration" section of the [Database Settings](/dashboard/project/_/settings/database):
```

![Connection Info and Certificate.](/docs/img/database/pool-size.png)

> The general rule is that if you are heavily using the PostgREST database API, you should be conscientious about raising your pool size past 40%. Otherwise, you can commit 80% to the pool. This leaves adequate room for the Authentication server and other utilities.

Reviewer 2: I think we're calling it Data API right? Can still reference PostgREST.

```suggestion
The general rule is that if you are heavily using the Data API (PostgREST), you should be conscientious about raising your pool size past 40%. Otherwise, you can commit 80% to the pool. This leaves adequate room for the Authentication server and other utilities.
```

Are we sure we can make this recommendation? I'm not sure if it does leave adequate room and highly depends on the compute instance.

I did an analysis here a while back: https://www.notion.so/supabase/Self-serve-database-connection-management-02248df94cee438aa87ba33e43b4e904?pvs=4#8f2c4df699744348b64ffc91bde3d434

Author: The suggestion is based on a [comment](https://supabase.slack.com/archives/C04JR9DBNQL/p1703625889822769?thread_ts=1703625707.918479&cid=C04JR9DBNQL) from @chasers. We can omit the suggestion, but I've found that it leads to indecision by developers. They need a concrete starting place. 

PostgREST uses a dynamic pooler that closes idle connections after 30s and will pend queries for up to 10s. Auth doesn't have a pooler and instead immediately closes connections after the request completes. 

Unless someone has sustained peaks in connection usage, it's unlikely that a collision will occur, even if they do overcommit. 

We can be more conservative with the recommendation and change to 70%

> These numbers are generalizations and assume a certain level of activity from all connected servers. The actual values depend on your peak connection usage. For instance, if you were only using 80 connections in a week period and your database could support 500 connections, then realistically you could allocate the remaining 420 (minus a reasonable buffer) to service more demand.

Reviewer 2: ```suggestion
These numbers are generalizations and depends on other Supabase products that you use and the extent of their usage. The actual values depend on your peak connection usage. For instance, if you were only using 80 connections in a week period and your database max connections is set to 500, then realistically you could allocate the difference of 420 (minus a reasonable buffer) to service more demand.
```

| ---------------- | ------------------------------------------------- |
| connection_id    | connection id                                     |
| ssl              | Indicates if SSL is in use                        |
| database         | Name of the connected database (usually postgres) |
| usename          | Role of the connected user                        |
| application_name | Name of the connecting application                |
| client_addr      | IP address of the connecting server               |
| query            | Last query executed by the connection             |
| query_start      | Time when the last query was executed             |
| state            | Querying state: active or idle                    |
| backend_start    | Timestamp of the connection's establishment       |

The usename can be used to identify the source:

| Role                         | API/Tool                                                                  |
| ---------------------------- | ------------------------------------------------------------------------- |
| supabase_admin               | Used by Supabase to configure projects and for monitoring                 |

> | authenticator                | PostgREST (Database API)                                                  |

Reviewer 2: ```suggestion
| authenticator                | Data API (PostgREST)                                                  |
```

> ---

Reviewer 2: @TheOtherBrian1 i think we should split this into two, one about Supavisor/Pooler and another one about monitoring connections.

feels a bit weird right now where the first half is all about supavisor and then have grafana/monitoring come in the picture.

Author: I considered adding the `monitoring connections` docs into 'connecting-to-postgres'. That would reduce the need for repetition and refreshers between the pages. However, I would like there to be some distance between the two. 

Monitoring connections is for three things:
- debugging connection bottlenecks
- determining optimal compute size
- adjusting pool size when using multiple APIs

These things aren't relevant to those sampling Supabase for the first time. Placing it upfront, creates needless noise and may even make users concerned about management burdens. 

What are your thoughts of shrinking the following passage and just moving it into the serverless drivers docs:
```
You can change how many database connections Supavisor can manage by altering the pool size in the [Database Settings](/dashboard/project/_/settings/database):

![Connection Info and Certificate.](/docs/img/database/pool-size.png)

The general rule is that if you are heavily using the PostgREST database API, you should be conscientious about raising your pool size past 40%. Otherwise, you can commit 80% to the pool. This leaves adequate room for the Authentication server and other utilities.

These numbers are generalizations and assume a certain level of activity from all connected servers. The actual values depend on your peak connection usage. For instance, if you were only using 80 connections in a week period and your database could support 500 connections, then realistically you could allocate the remaining 420 (minus a reasonable buffer) to service more demand.
```

Reviewer 2: let's just keep it in here for now

id: 'connection-management'
title: 'Connection management'
description: 'Managing connections'
subtitle: 'Using your connections resourcefully'
---

## Connections

> Every [Compute Add-On](https://supabase.com/docs/guides/platform/compute-add-ons) has a pre-configured direct connection count and Supavisor pool size. This guide discusses ways to observe and manage them resourcefully.

Reviewer 3: ```suggestion
Every [Compute Add-On](/docs/guides/platform/compute-add-ons) has a pre-configured direct connection count and Supavisor pool size. This guide discusses ways to observe and manage them resourcefully.
```

> ## What is a pool

Reviewer 2: I would rename this to something like "What is a database poo?l" or "Why is a database pool important?".

Author: I decided to take the section out.

When clients (backend servers) connect to PostgreSQL or Supavisor, they use the PostgreSQL Wire Protocol. Unlike HTTP requests, these connections are stateful and persistent, involving a costly startup but becoming efficient once established.

> Clients may create several active connections with Postgres, reducing the overhead of establishing connections between queries.

Reviewer 2: This could easily apply to both direct database connections and database connection pool. I would just focus on differentiating how a pool is useful etc.

Author: I decided to remove it. This was discussed thoroughly in the connecting-to-postgres docs. Because it was placed away from the primary docs, I thought it would be prudent to add a bit of a refresher.

## The pools in Supabase

> Supabase servers only create connections when needed, closing idle ones to conserve resources. If you're not using a service, no connections will be made. They are configured to not over-consume the database connections.

Reviewer 2: I recommend pulling this out completely b/c it's confusing and doesn't really help me understand Supavisor in any meaningful way.

Author: The section was originally much longer, discussing connection management for PostgREST, Realtime, etc. However, after talking to Steve, I was discouraged from adding it. Nobody wanted users to toy around with already optimized configs.

This portion was a remnant of the original section and was intended to minimize connection management anxiety around the Supabase Client. Without the extra context, I guess it seems out of place. I'll remove it

## Understanding Supavisor

Supavisor is used to connect with external libraries, especially for serverless deployments. In transaction mode, it acts as a database connection load balancer between your application and Postgres. This setup allows your clients (application servers) to share database connections more efficiently.

In serverless or auto-scaling environments, Supavisor helps manage a large number of client connections by optimizing their use of available database connections.

While Supavisor can handle an impressive amount of client connections, they are ultimately routed through a smaller pool of actual database connections, based on the capacity of your Postgres database.

### Configuring Supavisor's pool size

> You can change how many database connections Supavisor can manage by altering the pool size in the [Database Settings](https://supabase.com/dashboard/project/_/settings/database):

Reviewer 3: ```suggestion
You can change how many database connections Supavisor can manage by altering the pool size in the [Database Settings](/dashboard/project/_/settings/database):
```

> ![Connection Info and Certificate.](/docs/img/database/pool-size.png)

Reviewer 3: Is this the right image? (looks like the link 404s)

Author: Oops, the images were added to a different branch. Just added to this one.

The general rule is that if you are heavily using the PostgREST database API, you should be conscientious about raising your pool size past 40%. Otherwise, you can commit 80% to the pool. This leaves adequate room for the Authentication server and other utilities.

These numbers are generalizations and assume a certain level of activity from all connected servers. The actual values depend on your peak connection usage. For instance, if you were only using 80 connections in a week period and your database could support 500 connections, then realistically you could allocate the remaining 420 (minus a reasonable buffer) to service more demand.

## Monitoring connections

### Capturing historical usage

Supabase offers a Grafana Dashboard that records and visualizes over 200 project metrics, including connections. For setup instructions, check the [metrics docs](/docs/guides/platform/metrics).

Its "Client Connections" graph displays connections for both Supavisor and Postgres

> ![client connection graph](/docs/img/database/client-connections.png)

Reviewer 3: Image here also 404s

Author: Hmm, not really sure how I left these out. They should be present now
